src/KeyringController.ts

```diff
@@ -984,16 +985,42 @@ class KeyringController extends EventEmitter {
    * Clear Keyrings
    *
    * Deallocates all currently managed keyrings and accounts.
-   * Used before initializing a new vault.
+   * Used before initializing a new vault and after locking
+   * MetaMask.
    */
-  async #clearKeyrings(): Promise<void> {
+  #clearKeyrings() {
     // clear keyrings from memory
-    this.keyrings = [];
+    while (this.keyrings.length > 0) {
+      const keyring = this.keyrings.pop();
+      if (keyring) {
+        this.#disposeKeyring(keyring);
+      }
+    }
     this.memStore.updateState({
       keyrings: [],
     });
   }
 
+  /**
+   * Dispose Keyring
+   *
+   * The Trezor Keyring has a method called `dispose` that removes the
+   * iframe from the DOM. The Ledger Keyring has a method called `destroy`
+   * that clears the keyring event listeners. This method checks if the provided
+   * keyring has either of these methods and calls them if they exist.
+   *
+   * @param keyring - The keyring to dispose or destroy.
+   */
+  #disposeKeyring(keyring: Keyring<Json>) {
```

Comment:
This should probably be async; ideally all of the keyring API methods would be async so that they can be more easily implemented by snaps